feature: Support for snapped versions of Firefox and Chromium

vasilev opened the pull request:
In response to Canonical's high-pressure selling of Snapcraft.

## Snap in brief

Snap applies auto-generated AppArmor profiles located in `/var/lib/snapd/apparmor/profiles/`. [Profile template](https://github.com/snapcore/snapd/blob/release/2.57/interfaces/apparmor/template.go#L380-L383) and [data to populate the template](https://github.com/snapcore/snapd/blob/release/2.57/interfaces/apparmor/template_vars.go#L37).

These profiles restrict access to `$XDG_RUNTIME_DIR` (as reported in https://github.com/graysky2/profile-sync-daemon/issues/247), and at the same time allow access to specific `$XDG_RUNTIME_DIR`'s subdirectories, so `$XDG_RUNTIME_DIR/snap.firefox/` is accessible to snapped Firefox, and `$XDG_RUNTIME_DIR/snap.chromium/` is accessible to snapped Chromium.

Browser profiles are located in `~/snap/$SNAP_INSTANCE_NAME/common/` (`~/snap/firefox/common/` and `~/snap/chromium/common/` respectively).

## Possible solution

To comply with AppArmor restrictions mentioned above the in-memory part needs to be rerouted from the root of `$VOLATILE` to respective subdirectories of it.


Tested on Ubuntu 22.04.1 .

vasilev commented:
@graysky2, could you please take a look?

vasilev commented:
Converted to draft. Work in progress.

This solution needs more testing due to possible race conditions during system startup. PSD service may start before Snap-related services, so `$XDG_RUNTIME_DIR/snap.$SNAP_INSTANCE_NAME/` directories may not exist at the time of PSD's `startup` phase.

Possible solutions (at the time being):
1. Quick&Dirty: employ `--mkpath` Rsync option at some set of conditions.
2. Fine-tune system service ordering and startup conditions.

yutayu commented:
According to this article , https://gihyo.jp/admin/serial/01/ubuntu-recipe/0714
snap doesn't regcognize other than snap folder, so snap rejects folders like  /dev/shm.
I made tmpfs profile folder in snap firefox folder and copy profile in it , and it works.

This is maybe hint.

vasilev committed:
feature: Support for snapped versions of Firefox and Chromium.

As AppArmor profiles allow access only to `$XDG_RUNTIME_DIR/snap.$SNAP_INSTANCE_NAME/`, re-routed profiles of snapped browsers exactly into respective `$XDG_RUNTIME_DIR/snap.$SNAP_INSTANCE_NAME/` subdirectories.
